aten/src/ATen/native/ForeachOps.cpp

```diff
@@ -0,0 +1,22 @@
+
+#include <ATen/native/ForeachOps.h>
+#include <ATen/native/DispatchStub.h>
+
+namespace at {
+namespace native {
+
+std::vector<Tensor> _foreach_add(TensorList tensors, Scalar scalar) {
+  if (tensors.size() == 0) {
+    throw std::runtime_error("Tensor list can't be empty.");
+  }
+
```

Comment:
What is a good way to check for memory overlap without TensorIterator?

Reply:
TensorIterator ultimately calls one of the functions [in this file](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/MemoryOverlap.h) to check for memory overlap. What sort of memory overlap are you trying to check?

Reply:
is_non_overlapping_and_dense for individual tensors.  For different tensors, afaik there isnt an easy way to check.  IMO size/strides checks should go in multi_tensor_apply, not the individual functions.

Reply:
is_non_overlapping_and_dense for individual tensors.  For different tensors, afaik there isnt an easy way to check.  IMO size/strides checks should go in multi_tensor_apply, not the individual functions.

Reply:
@mcarilli, there are a few problems with having those checks in multi_tensor_apply: 
1. In case of CPU path, we would have to duplicate those
2. Lets assume 2 cases: TensorList + scalar and TensorList + TensorList. For the TL+TL case, we need to check if both TLs are the same size (TL0[0].size == TL1[0].size, TL0[1].size == TL1[1].size, ...). Where should this check be? If in individual function, then we will have different checks in different places. If we put everything into multi_tensor_apply, we will have to add a lot of if/else conditions to figure out which case we are checking.
I would suggest keeping those checks here for now, and later, once we add more ops, refactor it to be reusable and better positioned.
What do you think?

Reply:
@zou3519, tbh im still figuring out what that really means in a given context. I will get back to this thread once i get it :)

Reply:
Yeah, as @mcarilli suggested - using `is_non_overlapping_and_dense`. i thought its a suggestion for a new method :) Added that with a test.

Reply:
Since there is no out= argument for these operators, I don't think any memory overlap checking is necessary.